src/crimson/net/ProtocolV2.cc

```diff
@@ -1769,11 +1785,31 @@ void ProtocolV2::notify_write()
 
 // WAIT state
 
-void ProtocolV2::execute_wait()
+void ProtocolV2::execute_wait(bool wait_max_or_incremental)
 {
-  // TODO not implemented
-  // trigger_state(state_t::WAIT, write_state_t::delay, false);
-  ceph_assert(false);
+  trigger_state(state_t::WAIT, write_state_t::delay, true);
+  if (socket) {
+    socket->shutdown();
+  }
+  exit_protocol = seastar::with_gate(pending_dispatch,
```

Comment:
When the connection is closing, we need to make sure the waiting (with backoff) is stopped properly using `pending_dispatch`. I'm still thinking of a way to unify `pending_dispatch` and `exit_protocol`, but currently they are separated because protocol executions can be overlapped.

Reply:
yeah. IMHO, we are misusing the side-effect of `pending_dispatch`. as it was introduced to ensure that a protocol is properly closed after all message being dispatched to dispatcher are drained.

if we are expecting the backoff to finish before closing the connection , why not just cancel the backoff?

Reply:
> IMHO, we are misusing the side-effect of pending_dispatch. as it was introduced to ensure that a protocol is properly closed after all message being dispatched to dispatcher are drained.

The per-connection gate was introduced to guard all the related dangling futures/continuations when closing a connection, because its memory would be freed afterwards. There were some comments about it, for example https://github.com/ceph/ceph/pull/24656#discussion_r227199245 and https://github.com/ceph/ceph/pull/25716#discussion_r245223438. So I think It is not only used for waiting messages or `Dispatcher` tasks drained.

> if we are expecting the backoff to finish before closing the connection , why not just cancel the backoff

REPLACING and CLOSING states are already cancelling the backoff (see current implementation), but they still need to wait for WAIT state aborted before going on, which not only includes the running `Timer` but also the follow-up `execute_connecting()` operation.
* For REPLACING, it is waiting `execution_done` for WAIT state aborted;
* And for CLOSING, it is waiting `pending_dispatch` for WAIT state aborted;

Reply:
@tchaikov I have to admit this is not a good implementation, and even buggy potentially. For example, when closing a `Connection`, it should not close the `pending_dispatch` gate directly, because it will also block some important tasks like dispatch reset. So CLOSING may need to wait for `execution_done` before closing `pending_dispatch`. I'll have a fix later with UT implemented to check reset are correctly called.

Reply:
ack. sorry, it's just difficult to me to get my head around this every time i read this =(

yeah, i agree we need to improve this for the sake of readability or correctness.

and i don't want to block this PR because of this. look forward to your fix with unittest!